crates/wasi-c2/src/file.rs

```diff
@@ -0,0 +1,246 @@
+    pub filetype: Filetype,
+    pub base_caps: FileCaps,
+    pub inheriting_caps: FileCaps,
+    pub flags: FdFlags,
+}
+
+impl WasiFile for cap_std::fs::File {
+    fn datasync(&self) -> Result<(), Error> {
+        self.sync_data()?;
+        Ok(())
+    }
+    fn sync(&self) -> Result<(), Error> {
+        self.sync_all()?;
+        Ok(())
+    }
+    fn get_filetype(&self) -> Result<Filetype, Error> {
+        let meta = self.metadata()?;
+        // cap-std's Metadata/FileType only offers booleans indicating whether a file is a directory,
+        // symlink, or regular file.
+        // Directories should be excluded by the type system.
+        if meta.is_file() {
+            Ok(Filetype::RegularFile)
+        } else {
+            Err(Error::Badf) // XXX idk what to do here
+        }
+    }
+    fn get_fdflags(&self) -> Result<FdFlags, Error> {
+        // XXX cap-std doesnt expose append, dsync, nonblock, rsync, sync
+        todo!()
+    }
+    fn get_oflags(&self) -> Result<OFlags, Error> {
+        // XXX what if it was opened append, async, nonblock...
+        let perms = self.metadata()?.permissions();
+        if perms.readonly() {
+            Ok(OFlags::RDONLY)
+        } else {
+            Ok(OFlags::RDWR)
+        }
+    }
+    fn set_oflags(&self, _flags: OFlags) -> Result<(), Error> {
+        // XXX cap-std::fs::Permissions does not export a constructor to build this out of
+        #[allow(unreachable_code)]
+        self.set_permissions(todo!())?;
+        Ok(())
+    }
+    fn get_filestat(&self) -> Result<Filestat, Error> {
+        // XXX cap-std does not expose every part of filestat
+        #![allow(unreachable_code, unused_variables)]
+        let meta = self.metadata()?;
+        Ok(Filestat {
+            device_id: todo!(),
+            inode: todo!(),
+            filetype: self.get_filetype()?,
+            nlink: todo!(),
+            size: meta.len(),
+            atim: meta.accessed()?.into_std(),
+            mtim: meta.modified()?.into_std(),
+            ctim: meta.created()?.into_std(),
+        })
+    }
+    fn set_filestat_times(
+        &self,
+        _atim: Option<FilestatSetTime>,
+        _mtim: Option<FilestatSetTime>,
+    ) -> Result<(), Error> {
+        // XXX cap-std does not expose a way to set accessed time or modified time
+        todo!()
+    }
+    fn set_filestat_size(&self, size: u64) -> Result<(), Error> {
+        self.set_len(size)?;
+        Ok(())
+    }
+    fn allocate(&self, offset: u64, len: u64) -> Result<(), Error> {
+        let metadata = self.metadata()?;
+        let current_size = metadata.len();
+        let wanted_size = offset.checked_add(len).ok_or(Error::TooBig)?;
+        // This check will be unnecessary when rust-lang/rust#63326 is fixed
+        if wanted_size > i64::max_value() as u64 {
+            return Err(Error::TooBig);
+        }
+        if wanted_size > current_size {
+            self.set_len(wanted_size)?;
+        }
+        Ok(())
```

Comment:
system-interface 0.1.0 now has [`FileIoExt::allocate`](https://docs.rs/system-interface/0.1.0/system_interface/fs/trait.FileIoExt.html#tymethod.allocate) which does this.